components/product/variant-selector.tsx

```diff
@@ -74,15 +74,36 @@ export function VariantSelector({
           );
 
           // The option is active if it's in the url params.
-          const isActive = searchParams.get(optionNameLowerCase) === value;
+          const isActive = optimisticOptions.get(optionNameLowerCase) === value;
 
           return (
             <button
               key={value}
               aria-disabled={!isAvailableForSale}
               disabled={!isAvailableForSale}
               onClick={() => {
-                router.replace(optionUrl, { scroll: false });
+                startTransition(() => {
+                  const newOptimisticVariants = optimisticVariants.map((variant) => {
+                    const updatedOptions = variant.selectedOptions.map((option) => {
+                      if (option.name.toLowerCase() === optionNameLowerCase) {
+                        return { ...option, value: value };
+                      }
+                      return option;
+                    });
+
+                    return { ...variant, selectedOptions: updatedOptions };
+                  });
+
+                  optimisticOptions.set(optionNameLowerCase, value);
+
+                  setOptimsticVariants(newOptimisticVariants);
+                  setOptimisticOptions(new URLSearchParams(optimisticOptions.toString()));
+
+                  const optionUrl = createUrl(pathname, optimisticOptions);
+
+                  // Navigate without page reload
```

Comment:
I'm not sure this comment's 100% accurate. It seems that the navigation can't complete without a full page render – at least in the sense that the search params are up-to-date, etc. While the UI is ready to interact with due to the useOptimistic hook, it's still waiting for the full page render before the search params update, so it's now open to race conditions for anything reading from the search params (eg, adding to cart)